src/Raven.Server/Documents/Indexes/Static/TimeSeries/MapTimeSeriesIndex.cs

```diff
@@ -0,0 +1,213 @@
+        }
+
+        protected override IndexItem GetTombstoneByEtag(DocumentsOperationContext databaseContext, long etag)
+        {
+            throw new NotSupportedException("We do not process tombstones for TimeSeries");
+        }
+
+        protected override bool HasTombstonesWithEtagGreaterThanStartAndLowerThanOrEqualToEnd(DocumentsOperationContext databaseContext, string collection, long start, long end)
+        {
+            throw new NotSupportedException("We do not process tombstones for TimeSeries");
+        }
+
+        protected override bool IsStale(DocumentsOperationContext databaseContext, TransactionOperationContext indexContext, long? cutoff = null, long? referenceCutoff = null, List<string> stalenessReasons = null)
+        {
+            var isStale = base.IsStale(databaseContext, indexContext, cutoff, referenceCutoff, stalenessReasons);
+            if (isStale && stalenessReasons == null || _referencedCollections.Count == 0)
+                return isStale;
+
+            return StaticIndexHelper.IsStaleDueToReferences(this, databaseContext, indexContext, referenceCutoff, stalenessReasons) || isStale;
+        }
+
+        public override Dictionary<string, HashSet<CollectionName>> GetReferencedCollections()
+        {
+            return _compiled.ReferencedCollections;
+        }
+
+        protected override unsafe long CalculateIndexEtag(DocumentsOperationContext documentsContext, TransactionOperationContext indexContext,
+            QueryMetadata query, bool isStale)
+        {
+            if (_referencedCollections.Count == 0)
+                return base.CalculateIndexEtag(documentsContext, indexContext, query, isStale);
+
+            var minLength = MinimumSizeForCalculateIndexEtagLength(query);
+            var length = minLength +
+                         sizeof(long) * 2 * (Collections.Count * _referencedCollections.Count); // last referenced collection etags and last processed reference collection etags
+
+            var indexEtagBytes = stackalloc byte[length];
+
+            CalculateIndexEtagInternal(indexEtagBytes, isStale, State, documentsContext, indexContext);
+            UseAllDocumentsCounterCmpXchgAndTimeSeriesEtags(documentsContext, query, length, indexEtagBytes);
+
+            var writePos = indexEtagBytes + minLength;
+
+            return StaticIndexHelper.CalculateIndexEtag(this, length, indexEtagBytes, writePos, documentsContext, indexContext);
+        }
+
+        public override long GetLastItemEtagInCollection(DocumentsOperationContext databaseContext, string collection)
+        {
+            if (collection == Constants.Documents.Collections.AllDocumentsCollection)
+                throw new InvalidOperationException("TODO ppekrol");
+
+            // TODO [ppekrol] implement this properly
+            var timeSeries = DocumentDatabase.DocumentsStorage.TimeSeriesStorage.GetTimeSeriesFrom(databaseContext, collection, 0)
```

Comment:
TODO: implement in TS schema an index and a method that gets this instantly

Reply:
done